CONTRIBUTING.md

```diff
@@ -354,6 +355,73 @@ run benchmarks via the `make benchmarks` command. In addition, Vector
 maintains a full [test hardness][urls.vector_test_harness] for complex
 end-to-end integration and performance testing.
 
+### Profiling
+
+If you're trying to improve Vector's performance (or understand why your change
+made it worse), profiling is a useful tool for seeing where time is being spent.
+
+While there are a bunch of useful profiling tools, a simple place to get started
+is with Linux's `perf`. Before getting started, you'll likely need to give
+yourself access to collect stats:
+
+```sh
+echo -1 | sudo tee /proc/sys/kernel/perf_event_paranoid
+```
+
+You'll also want to edit `Cargo.toml` and make sure that Vector is being built
+with debug symbols in release mode. This ensures that you'll get human-readable
+info in the eventual output:
+
+```toml
+[profile.release]
+debug = true
+```
+
+Then you can start up a release build of Vector with whatever config you're
+interested in profiling.
+
+```sh
+cargo run --release -- --config my_test_config.toml
+```
+
+Once it's started, use the `ps` tool (or equivalent) to make a note of its PID.
+We'll use this to tell `perf` which process we would like it to collect data
+about.
```

Comment:
Would it not be simpler to do a release build and run `perf record … -- target/release/vector --config my_test_config.toml`?

Reply:
That's part of the trick here :smile: 

You can totally run it that way, but it will collect perf data for the entire lifetime of the process. This includes startup, shutdown, and any idle time where the server is not under load. That makes it a little tricky to time things if you only want perf data from the server while it's under load.

From that perspective, it's a bit simpler to give perf the load-generating command instead. You end up with a more focused flamegraph and don't need to worry about the timing of your commands.

Reply:
I wondered if that might be the case, but it wasn't obvious to me that was the purpose of this three-step dance (run, ps, perf). Perhaps a word about that would be in order.

Reply:
Good point, will make that clearer.